### PR TITLE
Breakdown of jobs by state

### PR DESCRIPTION
**Pull request by egze**
I added navigation to the page and it makes it possible to filter jobs by state. For example I only want to observe jobs in `executing` state.

Also improved some small things in the list table:
* Removed columns to focus only on the most important: worker, args, attempts, queue, timestamp.
* Showing relative timestamp in words, instead of a timestamp.
* Offer to see 20, 50 or 100 items per page (default 20).

We can also make it all configurable.

oban_daashboard

### CODE REVIEW

**Review by evilmarty — CHANGES_REQUESTED** [review]
@egze I want to start off by acknowledging and appreciating all your effort. I'm extremely humbled that you find this project useful and have taken time to improve it. I think your additions are great and with some adjustments can get to a point where they can be included.

I very much like the filtering by state, and the job count for each, but not at the expense of losing the "all" view. I personally find seeing all my jobs in one view very helpful, especially during development and I think from a usability standpoint it means an additional click to inspect a job that isn't executing by default. The functionality is useful though and it is still worthwhile to have.

Changing the time from a date to relative can be very useful but I don't believe it should replace it. There are times that I find it useful to see the date of a job. Making it possible to alternate the time format would be fantastic feature. I would go one step further and make it possible to display in local time, in addition to UTC and relative.

My main concern is the inclusion of Timex. I don't have an issue with the library itself but what the library includes. As you can see that by including Timex it is also including a HTTP client, SSL certs, parsers, etc. The most worrying is tzdata, which by default will make requests to fetch the latest timezone data. Not all projects may need dependencies, nor want them. In regard to the tzdata library, you have to update your config to disable auto-downloading. Not everyone may be aware of this when including this and it is something to be careful with.

The inclusion of Timex is a blocker because of what it depends on. It's excessive for just working out relative time. Elixir 1.18 includes a new Duration struct, which may open up an alternative path to achieve relative time outside of Timex in future. Or it may be possible to leverage the browser and use [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat), but that may be tricky to include client side JS via Live Dashboard. Another option forward would be to not include Timex but simply check if it is present and if so then provide the ability to display time relatively.

I would like to reiterate that I truly appreciate your effort and skill with this proposal. If would like to split out the relative time and job filtering into separate PRs that will help expedite getting one feature in whilst the other continues to progress. Thanks again.

File: lib/oban/live_dashboard.ex [hunk 1]
```diff
@@ -4,43 +4,46 @@ defmodule Oban.LiveDashboard do
...
+    <h1 class="mb-3">Oban</h1>
+
+    <p>Filter jobs by state:</p>
+
+    <.live_nav_bar id="oban_states" page={@page} nav_param="job_state" style={:bar} extra_params={["nav"]}>
+      <:item :for={{job_state, count} <- @job_state_counts} name={job_state} label={job_state_label(job_state, count)} method="navigate">
+        <.live_table id="oban_jobs" limit={per_page_limits()} dom_id={"oban-jobs-#{job_state}"} page={@page} row_attrs={&row_attrs/1} row_fetcher={&fetch_jobs(&1, &2, job_state)} default_sort_by={@timestamp_field} title="" search={false}>
+          <:col :let={job} field={:worker} sortable={:desc}>
+            <p class="font-weight-bolder"><%= job.worker %></p>
+            <pre class="args font-weight-lighter text-muted"><%= truncate(inspect(job.args)) %></pre>
```
[the remark this answers is not in the document]
Comment (egze) [reply lib/oban/live_dashboard.ex:31]: The args are useful to me, when the args are dynamic - it's nice to quickly spot a specific value. Also I have ideas to implement search and then filtering by args becomes even more sense.

I was inspired by the layout of Oban.Web and they have it https://getoban.pro/oban

File: lib/oban/live_dashboard.ex [hunk 2]
```diff
@@ -4,43 +4,46 @@ defmodule Oban.LiveDashboard do
...
-    <.live_modal
-      :if={@job != nil}
-      id="modal"
-      title="Job"
-      return_to={live_dashboard_path(@socket, @page, params: %{})}
-    >
+    <h1 class="mb-3">Oban</h1>
+
+    <p>Filter jobs by state:</p>
+
```
[the remark this answers is not in the document]
Comment (egze) [reply lib/oban/live_dashboard.ex:25]: I believe the concern is unnecessary. These are merely basic HTML tags, and there’s no indication from the LiveDashboard team that they discourage their use.

They even say that we can use any supported CSS classes https://github.com/phoenixframework/phoenix_live_dashboard/issues/438#issuecomment-1915125341

**Comment by egze** [comment]
> The inclusion of Timex is a blocker because of what it depends on. It's excessive for just working out relative time. Elixir 1.18 includes a new Duration struct, which may open up an alternative path to achieve relative time outside of Timex in future. Or it may be possible to leverage the browser and use [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat), but that may be tricky to include client side JS via Live Dashboard. Another option forward would be to not include Timex but simply check if it is present and if so then provide the ability to display time relatively.

I tried to not use Timex, and reach for JS. But unfortunately it's not cleanly possible at the moment. See discussion: https://github.com/phoenixframework/phoenix_live_dashboard/issues/455

What I can offer is: implement the relative_time feature without the Timex dependency. Someone already did it: https://gist.github.com/tlemens/88e9b08f62150ba6082f478a4a03ac52

And also we can make it configurable.

```elixir
additional_pages: [
  oban: {Oban.LiveDashboard, relative_timestamp: true}
]
```

If `false` then it would display it as today.

What do you think?

**Comment by egze** [comment]
> I very much like the filtering by state, and the job count for each, but not at the expense of losing the "all" view. I personally find seeing all my jobs in one view very helpful, especially during development and I think from a usability standpoint it means an additional click to inspect a job that isn't executing by default. The functionality is useful though and it is still worthwhile to have.

That makes sense. I'll bring the `All` tab back and make it the first one.

File: lib/oban/live_dashboard.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (egze) [reply lib/oban/live_dashboard.ex:31]: > not the intention of this dashboard, it's to give a simple overview of jobs

I see it a bit differently. I would like it to have the same feature set as Oban Web. Some missing things are: search, ability to retry/execute jobs, charts.

Please have a think about this direction. I would love to contribute and eventually reach this state with your library. But would totally get if you want to keep it simple.

File: lib/oban/live_dashboard.ex [hunk 3]
```diff
@@ -4,43 +4,46 @@ defmodule Oban.LiveDashboard do
...
+      <:item :for={{job_state, count} <- @job_state_counts} name={job_state} label={job_state_label(job_state, count)} method="navigate">
+        <.live_table id="oban_jobs" limit={per_page_limits()} dom_id={"oban-jobs-#{job_state}"} page={@page} row_attrs={&row_attrs/1} row_fetcher={&fetch_jobs(&1, &2, job_state)} default_sort_by={@timestamp_field} title="" search={false}>
+          <:col :let={job} field={:worker} sortable={:desc}>
+            <p class="font-weight-bolder"><%= job.worker %></p>
+            <pre class="args font-weight-lighter text-muted"><%= truncate(inspect(job.args)) %></pre>
+          </:col>
+          <:col :let={job} field={:attempt} header="Attempt" sortable={:desc}>
+            <span class="attempts font-weight-lighter">
+              <%= job.attempt %>/<%= job.max_attempts %>
+            </span>
```
[the remark this answers is not in the document]
Comment (egze) [reply lib/oban/live_dashboard.ex]: Removed the styles for `Attempts`.

But for `worker` and `args` I think it does make sense to have different styles.

**Comment by egze** [comment]
@evilmarty 
I released my own package to experiment more faster and see where it takes me https://github.com/egze/obanalyze

But I will also do my best to work on the current PR if you are still interested. Please check if there's still something you want to change.

File: lib/oban/live_dashboard.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (evilmarty) [reply lib/oban/live_dashboard.ex:25]: I have to admit, I was not aware that Live Dashboard was using Bootstrap. In that case, I think it's safe to use arbitrary tags that are styled by it.

A slight adjustment to make it more inline with the other pages. Firstly, the title should match. Secondly, I don't think it is necessary to include a label for the filtering. I may be subjective but it does seem out of place with a nav bar.
```suggestion
    <h5 class="mb-3">Oban</h5>
```

File: lib/oban/live_dashboard.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (evilmarty) [reply lib/oban/live_dashboard.ex]: Thinking some more on this, I have warmed to the idea of include the args along side the worker name. I do agree it can be useful.

> I see it a bit differently. I would like it to have the same feature set as Oban Web.

If you want those features then there is a solution already available. The reason why I created this alternative was because there was no solution to assist during development, to inspect and debug background jobs. The people who develop Oban Web fund the development of Oban and I have no intention of using this to take away someone's lunch, so to speak.

File: lib/oban/live_dashboard.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (evilmarty) [reply lib/oban/live_dashboard.ex]: Thanks. I agree with this now.

File: lib/oban/live_dashboard.ex [hunk 4]
```diff
@@ -4,43 +4,44 @@ defmodule Oban.LiveDashboard do
...
+    <h1 class="mb-3">Oban</h1>
+
+    <p>Filter jobs by state:</p>
+
+    <.live_nav_bar id="oban_states" page={@page} nav_param="job_state" style={:bar} extra_params={["nav"]}>
+      <:item :for={{job_state, count} <- @job_state_counts} name={job_state} label={job_state_label(job_state, count)} method="navigate">
+        <.live_table id="oban_jobs" limit={per_page_limits()} dom_id={"oban-jobs-#{job_state}"} page={@page} row_attrs={&row_attrs/1} row_fetcher={&fetch_jobs(&1, &2, job_state)} default_sort_by={@timestamp_field} title="" search={false}>
+          <:col :let={job} field={:worker} sortable={:desc}>
+            <p class="font-weight-bold"><%= job.worker %></p>
+            <pre class="font-weight-lighter text-muted"><%= truncate(inspect(job.args)) %></pre>
```
Comment (evilmarty) [inline lib/oban/live_dashboard.ex:31]: The margins should be removed to keep the content condensed otherwise when there's many records there is too much wasted space.
```suggestion
            <p class="font-weight-bold m-0"><%= job.worker %></p>
            <pre class="font-weight-lighter text-muted m-0"><%= truncate(inspect(job.args)) %></pre>
```

File: lib/oban/live_dashboard.ex [hunk 5]
```diff
@@ -91,28 +113,81 @@ defmodule Oban.LiveDashboard do
...
+  end
+
+  defp jobs_count_query("all" = _job_state) do
+    Oban.Job
+  end
+
+  defp jobs_count_query(job_state) do
+    Oban.Job
+    |> where([job], job.state == ^job_state)
+  end
```
Comment (evilmarty) [inline lib/oban/live_dashboard.ex:189]: There is a lot of duplication here. It would be better to consolidate them so that they can be reused more effectively.

```suggestion
  defp jobs_query(params, job_state) do
    Oban.Job
    |> filter_by_job_state(job_state)
    |> filter_by_params(params)
  end
  
  defp jobs_count_query("all") do
    Oban.Job
  end

  defp jobs_count_query(job_state) do
    filter_by_job_state(Oban.Job, job_state)
  end
  
  defp filter_by_params(queryable, %{sort_by: sort_by, sort_dir: sort_dir, limit: limit}) do
    queryable
    |> limit(^limit)
    |> order_by({^sort_dir, ^sort_by})
  end
  
  defp filter_by_job_state(queryable, job_state) do
    where(queryable, [job], job.state == ^job_state)
  end
```

File: mix.lock
Comment (evilmarty) [inline mix.lock:1]: Correct me if I am wrong but I don't see your changes using anything particular from these changed dependencies. If you're not then please remove them. Dependency changes should be done in a separate PR to minimise delta changes and it also makes it easier to update the changelog based on PRs.

File: test/oban/live_dashboard_test.exs
Comment (evilmarty) [inline test/oban/live_dashboard_test.exs:1]: Please add tests for filtering between states.

File: lib/oban/live_dashboard.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (egze) [reply lib/oban/live_dashboard.ex]: Oban Web is [going open source](https://podcast.thinkingelixir.com/227), so it's not taking anyones lunch. And I would use it actually, if it worked with SQLite. But only Oban works with it. Web and Pro have some things that only work with Postgres.

File: mix.lock
Comment (egze) [reply mix.lock:1]: Oops. Don't know how it got there. Will revert.

File: test/oban/live_dashboard_test.exs
Comment (egze) [reply test/oban/live_dashboard_test.exs:1]: Done a68eec2

I noticed that I had to do a `Oban.LiveDashboardTest.Repo.delete_all(Oban.Job)` in the test, because somehow the values from other tests were still there. Maybe something that you can look into later. Ideally the queries should be rolled back after each test.

Comment (evilmarty) [reply test/oban/live_dashboard_test.exs:1]: Thanks for the heads up. I'll look into proper tear downs.

File: test/oban/live_dashboard_test.exs [hunk 6]
```diff
@@ -14,23 +14,75 @@ defmodule Oban.LiveDashboardTest do
...
-  defp job_fixture(args \\ %{}) do
-    {:ok, job} = Oban.Job.new(args, worker: "FakeWorker") |> Oban.insert()
+  defp job_fixture(args \\ %{}, state \\ "executing") do
+    {:ok, job} =
+      Oban.Job.new(args,
+        worker: "FakeWorker",
+        state: state,
+        attempted_at: DateTime.utc_now()
+      )
+      |> Oban.insert()
```
Comment (evilmarty) [inline test/oban/live_dashboard_test.exs:84]: `attempted_at` is not present for all job states. Instead, it would be more appropriate to set those fields when required.
```suggestion
  defp job_fixture(args \\ %{}, opts \\ []) do
    opts  = Keyword.put(opts, :worker, "FakeWorker")
    {:ok, job} = Oban.Job.new(args, opts) |> Oban.insert()
```

**Review by evilmarty — CHANGES_REQUESTED** [review]
Thanks for make those changes. I have a made another round of comments and I feel comfortable to go ahead after they have been actioned.

**Comment by egze** [comment]
@evilmarty The new feature with the retry buttons caused some merge conflicts. I tried to fix it, but not sure anymore if I did it cleanly. Please test

File: lib/oban/live_dashboard.ex [hunk 7]
```diff
@@ -79,8 +71,29 @@ defmodule Oban.LiveDashboard do
     """
   end
 
+  @impl true
+  def mount(params, _, socket) do
+    socket =
+      socket
+      |> assign(job_state: Map.get(params, "job_state", "executing"))
+      |> assign(sort_by: Map.get(params, "job_state"))
```
Comment (evilmarty) [inline lib/oban/live_dashboard.ex:79]: This should be moved to `handle_params` otherwise `sort_by` and `job_state` won't get updated except for when the view is re-mounted.

**Review by evilmarty — CHANGES_REQUESTED** [review]

File: lib/oban/live_dashboard.ex (hunk 7 above)
Comment (egze) [reply lib/oban/live_dashboard.ex]: Done. `handle_params` deserves a refactor later IMO.

**Review by evilmarty — APPROVED** [review]
